crowbar_framework/lib/crowbar/lock.rb

```diff
@@ -0,0 +1,76 @@
+#
+# Copyright 2015, SUSE LINUX GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module Crowbar
+  class Lock
+    attr_accessor :logger
+    attr_accessor :path
+    attr_accessor :remote
+
+    def initialize(options = {})
+      @logger = options.fetch :logger, Rails.logger
+      @path = options.fetch :path, "/var/lock"
+      @remote = options.fetch :remote, false
+    end
+
+    def create(name, nodes = [])
+      if @remote
+        remote(name, nodes, action: "create")
+      else
+        local(name, action: "create")
+      end
+    end
+
+    def destroy(name, nodes = [])
+      if @remote
+        remote(name, nodes, action: "destroy")
+      else
+        local(name, action: "destroy")
+      end
+    end
+
+    def with_lock(name, nodes = [])
+      create(name, nodes)
+      yield
+    ensure
+      destroy(name, nodes)
+    end
+
+    protected
+
+    def remote(name, nodes, options = {})
+      action = options.fetch :action
+      if action == "create"
+        cmd, msg = "touch", "Creating"
```

Comment:
`touch` is not good enough; the file should contain info which uniquely identifies its creator.  Then `destroy` should check that it owns the file before removing.  For debugging purposes it would also be useful if the file contains human-readable info about why it was created (e.g. the barclamp / proposal name and revision number at least).